yetanotherco/aligned_layer
# refactor(batcher): retry function

**Motivation**

This PR implements the `retry_function()`, which receives a function to be retried using exponential backoff.

**Description**

Modifies the batcher to use `retry_function()` in the following calls:

- `user_balance_is_unlocked`
- `get_user_nonce_from_ethereum`
- `get_user_balance` 
- `get_gas_price`
- `s3::upload_object`
- `listen_new_blocks()`

Stops using our forked version of `ethers-rs` since is not longer needed.

**Observations**

The calls within the `Batcher::new()` function are not modified since in case of failure the program will panic (changes can be implemented if needed).
Also `create_new_task` was left for a different PR.

All modified calls use the following default parameters:

- `DEFAULT_MIN_DELAY  = 500`
- `DEFAULT_MAX_TIMES = 5`
- `DEFAULT_FACTOR = 2.0`

# How to Test

```
make anvil_start_with_block_time
make batcher_start_local
make batcher_send_burst_groth16
```

Now try killing the container running the s3 service. You should see 5 warnings separated by exponential time:
```
[2024-10-17T21:14:09Z WARN  aligned_batcher] Error uploading batch to s3 dispatch failure
```

- There are also some test in `batcher/aligned-batcher/src/retry.rs`

- You can also add a `sleep()` before the function you want to test in the batcher, and then kill Anvil before the call happens.
